File: backend/app/schemas/resource_optimization.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any


class ResourceRequirementSchema(BaseModel):
    unit_type: str
    required: int
    available: int
    selected: int


class FactorBreakdownSchema(BaseModel):
    eta_score: float
    route_safety_score: float
    ai_confidence_score: float
    capability_score: float
    availability_score: float
    operational_score: float


class CandidateUnitSchema(BaseModel):
    rescue_unit_id: str
    unit_code: str
    unit_name: str
    unit_type: str
    status: str
    optimization_score: float
    distance_meters: float
    estimated_duration_seconds: int
    average_confidence: float
    risk_level: str
    is_selected: bool
    selection_reason: str
    factor_breakdown: FactorBreakdownSchema
    route_geometry: Optional[Dict[str, Any]] = None


class SelectedUnitSchema(BaseModel):
    rescue_unit_id: str
    unit_code: str
    unit_name: str
    unit_type: str
    optimization_score: float
    distance_meters: float
    estimated_duration_seconds: int
    average_confidence: float
    risk_level: str
    selection_reason: str
    factor_breakdown: FactorBreakdownSchema
    route_geometry: Optional[Dict[str, Any]] = None


class ResourceOptimizationResponse(BaseModel):
    success: bool
    incident_id: str
    incident_code: Optional[str] = None
    resource_status: str
    required_resources: List[ResourceRequirementSchema]
    selected_units: List[SelectedUnitSchema]
    candidate_units: List[CandidateUnitSchema]
    optimization_summary: str
    analyzed_at: str


class OptimizedDispatchResponse(BaseModel):
    success: bool
    incident_id: str
    incident_code: Optional[str] = None
    dispatched_units_count: int
    dispatch_ids: List[str]
    dispatches: List[Dict[str, Any]]
    message: str

File: src/types/resourceOptimization.ts
export type ResourceOptimizationStatus =
  | 'OPTIMAL'
  | 'PARTIAL'
  | 'INSUFFICIENT_RESOURCES'
  | 'NO_SUITABLE_UNITS'
  | 'CONFLICT'
  | 'ALREADY_DISPATCHED';

export interface ResourceRequirement {
  unit_type: string;
  required: number;
  available: number;
  selected: number;
}

export interface FactorBreakdown {
  eta_score: number;
  route_safety_score: number;
  ai_confidence_score: number;
  capability_score: number;
  availability_score: number;
  operational_score: number;
}

export interface CandidateUnit {
  rescue_unit_id: string;
  unit_code: string;
  unit_name: string;
  unit_type: string;
  status: string;
  optimization_score: number;
  distance_meters: number;
  estimated_duration_seconds: number;
  average_confidence: number;
  risk_level: string;
  is_selected: boolean;
  selection_reason: string;
  factor_breakdown: FactorBreakdown;
  route_geometry?: {
    type: 'LineString';
    coordinates: [number, number][];
  };
}

export interface SelectedUnit {
  rescue_unit_id: string;
  unit_code: string;
  unit_name: string;
  unit_type: string;
  optimization_score: number;
  distance_meters: number;
  estimated_duration_seconds: number;
  average_confidence: number;
  risk_level: string;
  selection_reason: string;
  factor_breakdown: FactorBreakdown;
  route_geometry?: {
    type: 'LineString';
    coordinates: [number, number][];
  };
}

export interface ResourceOptimizationResponse {
  success: boolean;
  incident_id: string;
  incident_code?: string;
  resource_status: ResourceOptimizationStatus | string;
  required_resources: ResourceRequirement[];
  selected_units: SelectedUnit[];
  candidate_units: CandidateUnit[];
  optimization_summary: string;
  analyzed_at: string;
}

export interface OptimizedDispatchResponse {
  success: boolean;
  incident_id: string;
  incident_code?: string;
  dispatched_units_count: number;
  dispatch_ids: string[];
  dispatches: any[];
  message: string;
}
